publisher-sdk/src/main/java/com/criteo/publisher/logging/PublisherCodeSanitizer.kt

```diff
@@ -0,0 +1,263 @@
+      field = value
+      allowedStackCheck = if (value) {
+        regexBasedAllowedStackCheck
+      } else {
+        loopBasedAllowedStackCheck
+      }
+    }
+
+  fun sanitize(throwable: Throwable): Throwable {
+    return deepSanitize(throwable, mutableMapOf())
+  }
+
+  private fun deepSanitize(
+      original: Throwable,
+      visited: MutableMap<Throwable, Throwable>
+  ): Throwable {
+    // Handle circular reference of throwable. For instance, throwable1 is caused by throwable2 and vice versa
+    visited[original]?.let {
+      return it
+    }
+
+    val newException = if (original.isThrownByPublisher()) {
+      PublisherException()
+    } else {
+      if (isCloningThrowable) {
+        original.shallowCopy()
+      } else {
+        original
+      }
+    }
+
+    visited[original] = newException
+
+    sanitizeCause(original, newException, visited)
+    sanitizeSuppressedExceptions(original, newException, visited)
+    sanitizeStacktrace(original, newException)
+
+    return newException
+  }
+
+  private fun sanitizeCause(
+      original: Throwable,
+      newException: Throwable,
+      visited: MutableMap<Throwable, Throwable>
+  ) {
+    val originalCause = original.cause
+    if (originalCause != null && originalCause != original) {
+      with(throwableInternal) {
+        newException.realCause = deepSanitize(originalCause, visited)
+      }
+    }
+  }
+
+  @SuppressLint("NewApi") // Ok, minSdkLevel of the real SDK is 19, see DeviceUtil#isVersionSupported
+  private fun sanitizeSuppressedExceptions(
+      original: Throwable,
+      newException: Throwable,
+      visited: MutableMap<Throwable, Throwable>
+  ) {
+    val copySuppressed = original.suppressed.map {
+      deepSanitize(it, visited)
+    }
+
+    with(throwableInternal) {
+      newException.realSuppressedExceptions = copySuppressed
+    }
+  }
+
+  private fun sanitizeStacktrace(original: Throwable, newException: Throwable) {
+    val newStackTrace = mutableListOf<StackTraceElement>()
+    original.stackTrace.forEach {
+      if (it.isAllowed()) {
+        newStackTrace.add(it)
+      } else if (newStackTrace.isEmpty() || newStackTrace.last() != privateStackTraceElement) {
+        // Group the private stacks together
+        newStackTrace.add(privateStackTraceElement)
+      }
+    }
+
+    newException.stackTrace = newStackTrace.toTypedArray()
+  }
+
+  /**
+   * Copy the type and the message of the given exception.
+   *
+   * [Throwable] is [Serializable], then it is possible to deeply copy it via serialization trick.
+   * But an exception can be really big in memory because of:
+   * 1. many cause or suppressed exceptions
+   * 2. embedded data (such as an Android [android.app.Application])
+   *
+   * 1. is solved by only doing a shallow copy of the throwable.
+   * 2. is solved if this method is only called on throwable thrown by the SDK. Then the SDK has an implicit
+   * responsibility to not put lot of data in its throwable.
```

Comment:
By "lot of data", I mean to declare a non-static exception in an `Activity`/`Application`/`'View` for instance. In any way, this is considered as a bad practice and I hope that future maintainer will keep good ones.

Putting a limit is tricky: you'll need to add a bounded `OutputStream` which should fallback on something when limit is reached. But what is the good approach ? Throwing ? Stopping the log ?

Also please note that non-SDK throwables are replaced by `PublisherException` not to skip the copy, but to remove the class name and the throwable message. This is to prevent such issue:
```
// publisher code
throw new NameOfPublisherClassThatShouldNotLeak("user password is " + plainTextPassword)
```

Reply:
Finally I removed the copy part